core/controllers/base.py

```diff
@@ -341,9 +340,20 @@ def vaidate_and_normalize_args(self):
         payload_arg_keys = []
         request_arg_keys = []
         for arg in self.request.arguments():
-            if arg in self.ARGS_WHICH_DO_NOT_NEED_SCHEMA_VALIDATION:
+            if arg == 'csrf_token':
+                # 'csrf_token' has been already validated in the
+                # dispatch method.
                 continue
-            if arg == 'payload':
+            elif arg == 'source':
+                source_url = self.request.get('source')
+                regex_pattern = (
+                    r'http[s]?://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\(\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+' # pylint: disable=line-too-long
+                )
+                regex_verified_url = re.findall(regex_pattern, source_url)
+                if not regex_verified_url:
+                    raise self.InvalidInputException(
+                        'Not a valid source url.')
+            elif arg == 'payload':
                 payload_args = self.payload
                 if payload_args is not None:
                     payload_arg_keys = payload_args.keys()
```

Comment:
Do you still need the "else" at the bottom -- should it raise an exception instead if the code reaches there?

Reply:
Yes we still need it, because that part will accomodate the args which are not the part of payload but need schema validation (query string params). since we do not know the names of those args ahead of time, that else part will cover them. eg: version, username, image (examples are specific to some handlers)